### PR TITLE
Flekschas/ui for adding track source

### LINKED ISSUE #329: UI for adding/managing track source servers
95% of the time I modify a JSON config, it is for this purpose.

### PR DESCRIPTION
**Pull request by flekschas**
## Description

> What was changed in this pull request?

Added an interface to edit, add, and delete track sources. I also modernized the whole modal to be a multistep process that allows us to add more steps, i.e., for configuring a track before its added.

May-29-2019 16-51-37

> Why is it necessary?

Fixes #329 

## Checklist

- [x] Unit tests added or updated
- [x] Screenshot for visual changes (e.g. new tracks)
- [x] Updated CHANGELOG

### CODE REVIEW

File: app/index.html [hunk 1]
```diff
@@ -416,8 +416,8 @@
   { bounded: true },
 );
 
-hgApi.on('viewConfig', () => {
-  console.log('View Config changed');
+hgApi.on('viewConfig', (viewConfig) => {
+  console.log('View Config changed', JSON.parse(viewConfig).trackSourceServers);
```
[the remark this answers is not in the document]
Comment (flekschas) [reply app/index.html]: Good catch! No it's not. Will remove it.

**Comment by flekschas** [comment]
Please note that this needs more testing. The UI also needs improvements. Just wanted to put it out here so we can discuss / test it.

**Comment by pkerpedjiev** [comment]
Ok cool. 

**Things to test**

 - [ ] The list of tracks updates when a track source server is added
 - [ ] The list of tracks updates when a track source is deleted
 - [ ] The exported viewconf contains new new track source server
 - [ ] The exported viewconf doesn't contain deleted track source server

**Questions**

- Should it be possible to delete all track source server? 
     - [PK] I think yes. But we maybe we can have a 'restore default' button to restore `higlass.io/api/v1`. Or have it as a placeholder text

**Potential Improvements**

- It would be pretty cool if the clicking 'Save' automatically pinged the server to check if it's correct and added a checkmark next to it. Or even better, checked how many tracks the server contains and showed that next to the checkmark.

**Comment by flekschas** [comment]
I agree with everything and can at least manually confirm that

- The exported viewconf contains new new track source server
- The exported viewconf doesn't contain deleted track source server

work. Of course we need to add automatic tests for all the pieces.

Regarding the improvement I am wondering if the server provides the necessary endpoints? We could "duck type" and simple assume it's a higlass server when `/api/v1/tilesets/` returns `200`?

As much as I like the ideas for improvements I am wondering if we should try to split the base implementation and the improvements in 2 parts? I am not sure but I would assume that implementing this checking would require a bit of refactoring and ideally some new button components that conditionally render icons, change their color etc. (those could come in handy in other UI components once we have them). Any thoughts?

**Comment by pkerpedjiev** [comment]
Checking _could_ come later, but often times, things that are supposed to come later don't arrive at all.

The only thing son that list that I'd insist on are collapsing and tests. The Add Track dialog is already quite large.

With that said, those servers do have to be HiGlass servers, don't they?. And a call to server/api/v1/tilesets/ should indeed return 200 along with an object containing the number of tilesets: `{'count': 200, 'results': [...]}`.

**Comment by pkerpedjiev** [comment]
Oh, and naturally, @nvictus should weigh in.

**Comment by flekschas** [comment]
He also needs to review the code changes 😎

**Comment by nvictus** [comment]
Should the UI not appear above the track listing? It feels a bit awkward at the bottom.

**Comment by flekschas** [comment]
@nvictus Good point. I actually had the UI at the top before but I moved it to the button for the following reason: changing the track source server is most likely not the most used action when opening the track menu. I would assume most people either never do it because they use the base server or they do it once and then populate the view with tracks. Therefore, putting the UI at the top would take up quite a bit of space for a not-so-important action but usually the most important things are listed first.

I am thinking the track source UI should somehow be linked to the track list. Maybe it can go right below it or maybe there could be a cog-wheel icon attached next to the "Select tileset". Maybe we should sketch things out :)

**Comment by pkerpedjiev-zymergen** [comment]
Or just have it collapsed at the top.

**Comment by flekschas** [comment]
The UI is already collapsed by default but it's still wasting a lot of space. Google's search interface for example puts search options below the search bar because they are secondary.

screen shot 2018-11-29 at 9 38 55 am

Another option is of course to have a separate dialog only dedicated to adjusting the track source servers.

**Comment by flekschas** [comment]
This PR need to be refactored after #643 is ready & merged

**Comment by pkerpedjiev** [comment]
This is superseded by https://github.com/higlass/higlass/pull/649, right? Can we close this?

**Comment by flekschas** [comment]
It’s not superseded. It just needs to be updated after we switch to the new modal. The new modal uses the content to be displayed, just the wrapper changed.

**Comment by flekschas** [comment]
Aside, would you prefer to not have a UI for adding track source servers? I think in general this could be nice to have feature?

**Comment by pkerpedjiev** [comment]
How do you select multiple datasets? If I click on one it automatically goes to the track type selection dialog.

Jun-27-2019 13-42-08

**Comment by flekschas** [comment]
> How do you select multiple datasets? If I click on one it automatically goes to the track type selection dialog.
FYI, the gif above is outdated. This is the current version:

Jun-27-2019 17-44-25

**Comment by flekschas** [comment]
Closed due to its complexity and "unnecessity"

**Pull request closed without merging**